### ISSUE TITLE
Server on Linux and Software-Update

### ISSUE THREAD

**Issue by ajtak-angelus** [rzander/ruckzuck#95]
Hi, I would like to ask if it is possible to run RuckZuck server on Linux, according to the instructions:
https://www.c-sharpcorner.com/article/how-to-deploy-net-core-application-on-linux/
And your note: No Software-Update detection possible.
Why is this not possible?

Thank you!

**Comment by rzander** [comment]
Yes, .NET Core can run on Linux; I have a RuckZuck Proxy Server at home running on Ubuntu...

If you host a RuckZuck Proxy (https://github.com/rzander/ruckzuck/wiki/RuckZuck-Proxy-Server), it will forward the requests to detect updates to the main server. In this case, Update-detection will work but you can not interact or modify the logic..

If you want to host your own RuckZuck Server, you can use the plugin (RZ.SWLookup.Plugin.Azure), but you have to build and maintain your own "database" to detect the updates...

You may have to share your idea behind the question...

**Comment by ajtak-angelus** [comment]
Thank you for your application being so well written that it works under Linux.
I would like to have my own database with programs to be sure what programs are in the database.
Or how do you authenticate uploaded packages? To be at 101% that there will be no security issue? I haven't read the source code yet.
It is only a hypothetical question. I really appreciate your work!

**Comment by rzander** [comment]
There is no 101% guarantee...

What RuckZuck does:
- Software uploads requires approval before the product becomes public available
- Download Source must be the vendors web-site (or e.g. GitHub if SW is Open-Source)
- RuckZuck does validate the file hash or the X509 signature to guarantee the file integrity.
- Every Package will be tested on a device with a current Defender engine (Note: not every Alert is a Virus; e.g. Sysinternal tools may generate a PUA (potentially unwanted software) alert).
- Some software will be downloaded from a RuckZuck server if the original source is slow, unstable etc.

Even if you host you own Repository, you still have to trust the vendor...

**Comment by ajtak-angelus** [comment]
Thank you for the explanation. I like the principle of operation of the installation, where you only provide a configuration file - similar to AppGet (now owned by MS winget).
How is it verified that the json file has not been modified along the way?
What does an individual JSON digitally sign?

**Comment by rzander** [comment]
The communication with the REST API is SSL encrypted. It's always a decision between security and flexibility... If we would enforce signing the metadata, no one would be able to have custom packages or modifications (as you can do with a RuckZuck Proxy)..

**Comment by ajtak-angelus** [comment]
hmm, I understand. Is it possible to install a proxy without a docker? And the last question is how to use rzupdate (command line) to install multiple applications at once (rzupdate firefox thunderbird 7-zip ...)
Thank you for your work!

**Comment by rzander** [comment]
Yes, you can run RZ.Server without docker... I recommend to build the binaries with VisualStudio as the last published  RZ.Server is in Release 1.7.1.3 (https://github.com/rzander/ruckzuck/releases/tag/1.7.1.3) . just follow: https://github.com/rzander/ruckzuck/wiki/RuckZuck-Server-with-custom-Repository 

There seems to be an issue running rzupdate with powershell, you have to run `&rzupdate.exe '"firefox";"thunderbird";"7-zip"'`  (add a single quote over all parameters)
on cmd it would be : `rzupdate.exe "firefox";"thunderbird";"7-zip"`  

I will try to fix that...  But in general, I would recommend to use https://github.com/rzander/ruckzuck/wiki/RuckZuck-OneGet-Provider , if you are using PowerShell...

**Comment by ajtak-angelus** [comment]
Hello, again :)
I compiled RZ.Plugin.SWLookup.Azure.dll and placed it /wwwroot/plugins/ and removing all the proxy dlls. The default program /repository/zander tools /DevCDRAgentCore/1.0.0.33/DevCDRAgent.json is in the /wwwroot/repository folder.
However, the program is not available on ip: 5000 - server is on Linux. Can you please advise me. Thank you

**Comment by rzander** [comment]
When you start the server with `dotnet rz.server.dll` , it will tell you which port is used: e.g. `Now listening on: http://[::]:5000 `. You can change the port by setting an environment variable `WebPort` to a custom port you want.. 

Note: All Plugins ending with "Azure" will require an Azure Storage Account... The DLL without the right configuration and backend will not do anything.. Plugins ending with "Proxy" do forward incoming requests to the public ruckzuck service. All the other Plugins will access local resources.
E.g. to see the app `/repository/DevCDRAgent.json` in the Catalog, you need `RZ.Plugin.Catalog.dll' as this is a local resource... to install the app, you need `RZ.Plugin.Software.dll`
If you have the Plugins `RZ.Plugin.Catalog.dll' and `RZ.Plugin.Catalog.Proxy.dll' it will get the local catalog and merge it with the public ruckzuck catalog.. that's how you can override or modify the public catalog.

**Comment by ajtak-angelus** [comment]
the web access works, but I don't see any application in the repository. So I have these two plugins RZ.Plugin.Catalog.dll and RZ.Plugin.Software.dll. I don't want to use a proxy.
Thank you!

**Comment by ajtak-angelus** [comment]
Some advice? Thank you! :)

**Comment by rzander** [comment]
The example `wwwroot\repository\DevCDRAgent.json` does remove that app from the catalog. Copy the file wwwroot\repository\zander tools\DevCDRAgentCore\1.0.0.33\DevCDRAgent.json' to `wwwroot\repository\DevCDRAgent.json` and restart the server.
The check the catalog, call the REST API `Invoke-RestMethod -uri "https://localhost:44307/rest/v2/getcatalog"`

**Comment by ajtak-angelus** [comment]
yes, I know - I read the instructions. But when you run dotnet ./RZ.Server.dll, the webpage will display:

![image](https://user-images.githubusercontent.com/19210976/89020612-80687e00-d31f-11ea-8e6b-ea75eb0ecc06.png)

**Comment by rzander** [comment]
I've published the latest RZ.Server here: https://github.com/rzander/ruckzuck/releases/tag/1.7.1.5

**Comment by ajtak-angelus** [comment]
Thank you! The repository is now available. But applications cannot be installed.
![image](https://user-images.githubusercontent.com/19210976/89133516-bf175780-d51c-11ea-97ad-e60ccae91d1d.png)

**Comment by ajtak-angelus** [comment]
Any suggestion, please :)

**Comment by rzander** [comment]
Can you "Edit" this Package (right click->Edit) ? it has a requirement to run only on X64 systems...

**Comment by ajtak-angelus** [comment]
nothing happens after clicking on the edit button. Even if I try another package - for example Firefox.

**Comment by rzander** [comment]
I can't reproduce you problem... it's working here.
What I did:
- Downloaded the Server https://github.com/rzander/ruckzuck/releases/download/1.7.1.5/RZ.Server.zip
- Removed all "Proxy" Plugins
![image](https://user-images.githubusercontent.com/11909453/89705094-a6fb6a00-d95a-11ea-96c2-5c4ef02e8cd5.png)
- Copied the DevCDRAgent.json to the root of repository (keep the original File !)
![image](https://user-images.githubusercontent.com/11909453/89705120-c85c5600-d95a-11ea-8d25-688e7b23fab8.png)
- Started RZ.Server.exe
- Changed Registry `HKEY_LOCAL_MACHINE\SOFTWARE\Policies\RuckZuck` and set `WebService` to `http://localhost:5000`
- Deleted `%TEMP%\rzcat.json`
- Started RuckZuck.exe
![image](https://user-images.githubusercontent.com/11909453/89705155-27ba6600-d95b-11ea-9e94-55995208bb91.png)
- And I can edit:
![image](https://user-images.githubusercontent.com/11909453/89705165-343ebe80-d95b-11ea-91c2-101413938e53.png)

**Comment by ajtak-angelus** [comment]
RuckZuck server run on Linux Debian Buster. Otherwise, the procedure is the same. :(

**Comment by rzander** [comment]
Can you run your files on Windows? just to confirm that it's running under Windows... Catalog Plugin seems to work as you can see the Catalog.

**Comment by rzander** [comment]
Can you try to rename everything under `repository` to lowercase... I think linux is case sensitive and the Software plugin does convert everything to lowercase...

**Comment by ajtak-angelus** [comment]
BRAVO! Rename all to lowercase works!

**Comment by rzander** [comment]
:-) ok, I will update the example to lowercase all files and directories...

**Comment by ajtak-angelus** [comment]
Wouldn't it be better not to be case sensitive?

**Comment by rzander** [comment]
Is far as I understand it, it's the file system which is case sensitive... It may work with case sensitive folders as long as you use only the ruckzuck.exe to access the repository, but as soon as you start using the REST API, the OneGet Provider, RZGet.exe etc. you must now how the exact case to query software... The public repository on Azure Blob Storage is also case sensitive and all items are stored in lowercase.

**Comment by ajtak-angelus** [comment]
clear, I understand. I would like to ask more. If I copy RZ.Plugin.SWLookup.dll to the Plugins folder. Will it be possible to update applications?

**Comment by rzander** [comment]
Updated https://github.com/rzander/ruckzuck/releases/download/1.7.1.5/RZ.Server.zip including a readme.txt to remember that all folders and files must be lowercase.

RZ.Plugin.SWLookup.dll does not contain any logic to detect updates, it just returns a blank JSON array...
```C#
        public JArray CheckForUpdates(JArray Softwares, string customerid = "")
        {
            return new JArray();
        }
```
But you are free to extend it to detect updates...

**Comment by ajtak-angelus** [comment]
I have tested other applications. If I add VLC, it can be installed but not edited. Firefox can be edited.

├── mozilla
│   └── mozilla firefox 79.0 (x64)
│       └── 79.0
│           └── firefox.json
├── videolan
│   └── vlc media player
│       └── 3.0.11
│           └── vlc.json
├── zander tools
│   └── devcdragentcore
│       └── 1.0.0.33
│           └── devcdragent.json
├── devcdragent.json
├── firefox.json
└── vlc.json

RZ.Plugin.SWLookup.dll - good idea :)

**Comment by khanmira663** [comment]
this has been helpful 
thanks a lot guys for sharing such a informative discussion platform 
[download](https://forcrack.com/autocad-2017-crack/)
